### PR TITLE
[Internal] Added self-hosted runner in run system tests workflow

### PR DESCRIPTION
**Pull request by MounikaBattu17**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
- This PR splits `run_tests` workflow into `run_unit_tests` & `run_system_tests` workflows.
- **run_unit_tests**: 
   - runs all unit tests of `ni-measurementlink-service`  folder `./tests/unit` with and without `extras` install.
   - runs all tests of `ni-measurementlink-generator`  folder.
   - runs on GitHub provided hosts (macros-latest, windows-latest, ubuntu-latest).
- **run_system_tests**: 
   - runs all acceptance and integration tests of `ni-measurementlink-service`  folder `./tests/acceptance`, `./tests/integration`  with all `extras` installed as these tests require real drivers.
   - used a `tox` environment to run these tests on multiple python version to reduce the number of runners.
   - runs on self-hosted runner (`rdss-measlinkbot-win-10-py32`, `rdss-measlinkbot-win-10-py64`).

### Why should this Pull Request be merged?
Implements [Task 2534163](https://dev.azure.com/ni/DevCentral/_workitems/edit/2534163): Set up self hosted runner in repo and verify the repo pipeline
Completes [Technical Debt 2487132](https://dev.azure.com/ni/DevCentral/_workitems/edit/2487132): Need to have a self-hosted runner with MeasurementLink installed (GitHub Issue #351)

### What testing has been done?
Tested pipeline run system tests workflow.

### CODE REVIEW

File: .github/workflows/run_system_tests.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      - name: Check out repo
+        uses: actions/checkout@v3
+      - name: Install dependencies
+        run: poetry install
+      - name: Install tox
+        run: pip install tox
+      - name: Run system tests
+        run: poetry run tox
+        env:
+          passenv: RUNNERNAME
```
Comment (MounikaBattu17) [inline .github/workflows/run_system_tests.yml:35]: This logic of passing the runner name to tox is still in testing.

**Comment by MounikaBattu17** [comment]
Addressing Brad's comment from [here](https://github.com/ni/measurementlink-python/pull/476#discussion_r1372230911)




> I think it will be simpler to split run_test.yml into run_unit_tests.yml and run_system_tests.yml like we did in [nidaqmx-python](https://github.com/ni/nidaqmx-python/tree/master/.github/workflows).
> 
> Other things we may want to do differently for each one:
> 
> Timeout (see nidaqmx-python)
- Added timeout as 90 minutes
> We may not want to include run_system_tests.yml in the required status checks if it's slow/unreliable
- It is better to mark run_system_tests.yml as not required since we face some issues with py64 runner currently.
> nimg tests don't need any services/drivers, so we only need to include them in run_unit_tests.yml
> We may want to handle Python versions differently for run_system_tests.yml to reduce the number of runners
- Used tox environment approach to test multiple python versions. But one of the disadvantage of this is it takes longer time to run tests than for running tests on individual python versions.
> We can use pytest command line options (-k or -m) to filter out tests that require drivers/services in run_unit_tests.yml so they don't add to the skip count.
- yet to implement
> Make sure the test result files have unique filenames so report_test_results.yml will treat them as separate tests.
- I have added RUNNERNAME approach to pass runner name to tox.ini -> still in testing

**Comment by MounikaBattu17** [comment]
some observations on the runners:
- when running jobs on individual python versions, self-hosted runner picks up the jobs faster and executes but incase on single runner requirement like tox, it takes more time or never in picking up the job
- py64 runner is not running even the webhook is not triggered.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+isolated_build = true
+envlist = clean, py{38,39,310,311,32}-all-extras
+
+[testenv]
+skip_install = true
+allowlist_externals = poetry
+setenv =
+   passenv = RUNNERNAME
+   # setting this value to use system environment path.
+   ProgramW6432 = C:/Program Files
```
Comment (bkeryan) [inline tox.ini:16]: I don't understand why you're overriding this.

Comment (bkeryan) [reply tox.ini:16]: Is this error the reason you are setting ProgramW6432 or is it the result of you setting ProgramW6432?

```
E           KeyError: 'ProgramW6432'
C:\Program Files\Python39\lib\os.py:679: KeyError
_ ERROR at setup of test___multi_session___measure___reserves_multiple_sessions[configuration10] _
    @pytest.fixture(scope="session")
    def discovery_service_process() -> Generator[DiscoveryServiceProcess, None, None]:
        """Test fixture that creates discovery service process."""
        if sys.platform != "win32":
            pytest.skip(
                f"Platform {sys.platform} is not supported for starting discovery service."
                "Could not proceed to run the tests."
            )
>       elif not _get_registration_json_file_path().exists():
tests\conftest.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ni_measurementlink_service\discovery\_support.py:59: in _get_registration_json_file_path
    pathlib.Path(os.environ["ProgramW6432"])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = environ({'COMSPEC': 'C:\\Windows\\system32\\cmd.exe', 'NUMBER_OF_PROCESSORS': '2', 'PATHEXT': '.COM;.EXE;.BAT;.CMD;.VB...nce/test_session_management.py::test___multi_session___measure___reserves_multiple_sessions[configuration10] (setup)'})
key = 'ProgramW6432'
    def __getitem__(self, key):
        try:
            value = self._data[self.encodekey(key)]
        except KeyError:
            # raise KeyError with the original key value
```

Comment (bkeryan) [reply tox.ini:16]: Anyway, I logged into the runner and ran `dir env:`, and this variable was set, as expected. You should not have to set it.

File: .github/workflows/run_system_tests.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    steps:
+      - name: Set Runner Environment Variables
+        run: |
+          echo "RUNNERNAME=rdss-measlinkbot-win-10-${{ matrix.configuration }}" >> $GITHUB_ENV
+      - name: Check out repo
+        uses: actions/checkout@v3
+      - name: Install dependencies
+        run: poetry install
+      - name: Install tox
+        run: pip install tox
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply .github/workflows/run_system_tests.yml]: Added `tox = ">=3.24"` in `pyproject.toml` file.

File: .github/workflows/run_system_tests.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      - rdss-measlinkbot-win-10-${{ matrix.configuration }}
+    strategy:
+      matrix:
+        configuration: ["py32", "py64"]
+      # Fail-fast skews the pass/fail ratio and seems to make pytest produce
+      # incomplete JUnit XML results.
+      fail-fast: false
+    timeout-minutes: 90
+    steps:
+      - name: Set Runner Environment Variables
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply .github/workflows/run_system_tests.yml]: Updated.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Tox (http://tox.testrun.org/) is a tool for running tests
+# in multiple virtualenvs. This configuration file will run the
+# test suite on all supported python versions. To use it, "pip install tox"
+# and then run "tox" from this directory.
+
+[tox]
+isolated_build = true
+envlist = clean, py{38,39,310,311,32}-all-extras
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply tox.ini]: Corrected.

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+skip_install = true
+allowlist_externals = poetry
+setenv =
+   passenv = RUNNERNAME
+   # setting this value to use system environment path.
+   ProgramW6432 = C:/Program Files
+commands =
+   poetry run python --version
+   poetry install -v --all-extras
+   poetry run pytest ./tests/acceptance ./tests/integration -v --cov=ni_measurementlink_service --cov-append --cov-report= --junitxml=test_results/nims-{env:RUNNERNAME}-all_extras.xml
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply tox.ini]: Updated the `poetry run pytest` command such that it runs all tests.

[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply tox.ini]: Appended `{envname}` such the name goes `py38-all-extras`, `py39-all-extras` and so on.

File: tox.ini (hunk 2 above)
Comment (MounikaBattu17) [reply tox.ini]: Yes, I saw this error in pipeline that's why I overrided the key `ProgramW6432`. Now I have removed.

File: pyproject.toml [hunk 7]
```diff
@@ -1,9 +1,9 @@
 [tool.black]
-extend_exclude = '_pb2(_grpc)?\.(py|pyi)$|ni_measurementlink_generator/|examples/'
+extend_exclude = '.tox/,_pb2(_grpc)?\.(py|pyi)$|ni_measurementlink_generator/|examples/'
```
Comment (bkeryan) [inline pyproject.toml:2]: Black and ni-python-styleguide use different formats here. Black wants a single regex, and ni-python-styleguide wants a comma-separated list of fnmatch patterns. 😩

```suggestion
extend_exclude = '\.tox/|_pb2(_grpc)?\.(py|pyi)$|ni_measurementlink_generator/|examples/'
```

File: .github/workflows/run_unit_tests.yml [hunk 8]
```diff
@@ -13,7 +13,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [macos-latest, ubuntu-latest, windows-latest]
+        os: [macos-latest, windows-latest, ubuntu-latest]
```
Comment (bkeryan) [inline .github/workflows/run_unit_tests.yml]: Revert unnecessary change
```suggestion
        os: [macos-latest, ubuntu-latest, windows-latest]
```

File: pyproject.toml (hunk 7 above)
Comment (MounikaBattu17) [reply pyproject.toml]: Corrected.

File: poetry.lock [hunk 9]
```diff
@@ -1,1983 +1,2016 @@
...
-]
-urllib3 = [
-    {file = "urllib3-2.0.7-py3-none-any.whl", hash = "sha256:fdb6d215c776278489906c2f8916e6e7d4f5a9b602ccbcfdf7f016fc8da0596e"},
-    {file = "urllib3-2.0.7.tar.gz", hash = "sha256:c97dfde1f7bd43a71c8d2a58e369e9b2bf692d1334ea9f9cae55add7d0dd0f84"},
-]
-zipp = [
-    {file = "zipp-3.17.0-py3-none-any.whl", hash = "sha256:0e923e726174922dce09c53c59ad483ff7bbb8e572e00c7f7c46b88556409f31"},
-    {file = "zipp-3.17.0.tar.gz", hash = "sha256:84e64a1c28cf7e91ed2078bb8cc8c259cb19b76942096c8d7b84947690cabaf0"},
-]
+# This file is automatically @generated by Poetry 1.6.1 and should not be changed by hand.
```
Comment (bkeryan) [inline poetry.lock:1]: Sorry, this component is still using Poetry 1.2.2 for now.

File: .github/workflows/run_unit_tests.yml (hunk 8 above)
Comment (MounikaBattu17) [reply .github/workflows/run_unit_tests.yml]: Reverted back.

File: poetry.lock (hunk 9 above)
Comment (MounikaBattu17) [reply poetry.lock]: Installed poetry `1.2.2` version and generated `poetry.lock` file.

File: tox.ini (hunk 2 above)
Comment (MounikaBattu17) [reply tox.ini]: Seems like tox environment is not able to access system environment variables, so I have passed in `ProgramW6432` to tox environment. This fixed the above key error issue.
`passenv = RUNNER_NAME, ProgramW6432`

Comment (bkeryan) [reply tox.ini]: Thanks, that explains it. You may also need to pass `ProgramData`.

I posted a PR to remove the dependency on these environment variables: https://github.com/ni/measurementlink-python/pull/483

File: .github/workflows/run_unit_tests.yml [hunk 10]
```diff
@@ -8,12 +8,12 @@ env:
...
 jobs:
-  run_tests:
-    name: Run tests
+  run_unit_tests:
+    name: Run unit tests
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [macos-latest, windows-latest, ubuntu-latest]
+        os: [macos-latest, ubuntu-latest, windows-latest]
```
Comment (bkeryan) [inline .github/workflows/run_unit_tests.yml]: revert unnecessary change again

```suggestion
        os: [macos-latest, windows-latest, ubuntu-latest]
```

File: .github/workflows/run_unit_tests.yml (hunk 8 above)
Comment (bkeryan) [reply .github/workflows/run_unit_tests.yml]: Sorry, I didn't realize the diff I was looking at was already reverting this.

**Review by bkeryan — APPROVED** [review]

File: .github/workflows/run_unit_tests.yml (hunk 8 above)
Comment (bkeryan) [reply .github/workflows/run_unit_tests.yml]: I reverted this again.

File: tox.ini (hunk 2 above)
Comment (bkeryan) [reply tox.ini]: I added ProgramData to the list.
